File: src/runtime/glyphVM.ts
import type { ToolStatus } from "../harness/types.js";
import type { ToolRegistry } from "../harness/toolRegistry.js";
import { parseGlyphToIR, type GlyphIR, type GlyphIRStep, type GlyphIRValue, type GlyphRepairStep, type GlyphToolStep } from "../ir/glyphIR.js";
import { validateIR } from "../ir/validateIR.js";
import { RuntimeContext } from "./context.js";
import { GlyphRuntimeError } from "./errors.js";
import type { TraceEvent } from "./trace.js";

export interface GlyphVMRunResult {
  ir: GlyphIR;
  trace: TraceEvent[];
  outputs: unknown[];
  variables: Record<string, unknown>;
  mockFS: Record<string, unknown>;
}

export interface GlyphVMOptions {
  mockFS?: Record<string, unknown>;
}

export class GlyphVM {
  constructor(private readonly registry: ToolRegistry) {}

  async runSource(source: string, options: GlyphVMOptions = {}): Promise<GlyphVMRunResult> {
    const ir = validateIR(parseGlyphToIR(source));
    return this.execute(ir, options);
  }

  async execute(ir: GlyphIR, options: GlyphVMOptions = {}): Promise<GlyphVMRunResult> {
    const validIR = validateIR(ir);
    const ctx = new RuntimeContext({ context: validIR.context, mockFS: options.mockFS });

    for (const flow of validIR.flows) {
      for (const step of flow.steps) {
        await this.executeStep(step, ctx);
      }
    }

    return {
      ir: validIR,
      trace: ctx.trace.all(),
      outputs: [...ctx.outputs],
      variables: ctx.variables.snapshot(),
      mockFS: ctx.snapshotFS()
    };
  }

  private async executeStep(step: GlyphIRStep, ctx: RuntimeContext, iteration?: number): Promise<void> {
    if (step.kind === "tool") {
      await this.executeToolStep(step, ctx, iteration);
      return;
    }

    await this.executeRepairStep(step, ctx);
  }

  private async executeToolStep(step: GlyphToolStep, ctx: RuntimeContext, iteration?: number): Promise<void> {
    const started = Date.now();
    let resolvedArgs: Record<string, unknown> = {};

    try {
      resolvedArgs = this.resolveArgs(step.args, ctx, step.id);
      const tool = this.registry.get(step.op);

      if (!tool) {
        throw new GlyphRuntimeError(`Unknown tool "${step.op}"`, step.id);
      }

      const result = await tool(resolvedArgs, ctx);

      if (step.assignTo) {
        ctx.variables.set(step.assignTo, result.value);
      }

      if (step.op === "EXPORT") {
        ctx.outputs.push(result.value);
      }

      ctx.trace.add({
        stepId: step.id,
        operation: step.op,
        resolvedArgs,
        outputSummary: result.summary,
        status: result.status,
        durationMs: Date.now() - started,
        errors: result.status === "fail" ? result.warnings : undefined,
        iteration
      });
    } catch (error) {
      ctx.trace.add({
        stepId: step.id,
        operation: step.op,
        resolvedArgs,
        outputSummary: "Step failed",
        status: "fail",
        durationMs: Date.now() - started,
        errors: [error instanceof Error ? error.message : String(error)],
        iteration
      });
      throw error;
    }
  }

  private async executeRepairStep(step: GlyphRepairStep, ctx: RuntimeContext): Promise<void> {
    const started = Date.now();

    try {
      this.requireVariable(step.targetVar, ctx, step.id);
      this.requireVariable(step.reportVar, ctx, step.id);

      let iterations = 0;
      for (let index = 0; index < step.maxIterations; index += 1) {
        if (this.reportStatus(ctx.variables.get(step.reportVar)) === "pass") {
          break;
        }

        iterations = index + 1;
        for (const innerStep of step.steps) {
          await this.executeStep(innerStep, ctx, iterations);
        }

        if (this.reportStatus(ctx.variables.get(step.reportVar)) === "pass") {
          break;
        }
      }

      const finalStatus = this.reportStatus(ctx.variables.get(step.reportVar));
      ctx.trace.add({
        stepId: step.id,
        operation: "REPAIR",
        resolvedArgs: {
          target: ctx.variables.get(step.targetVar),
          report: ctx.variables.get(step.reportVar),
          max: step.maxIterations
        },
        outputSummary: `Repair loop completed after ${iterations} iteration${iterations === 1 ? "" : "s"}`,
        status: finalStatus,
        durationMs: Date.now() - started
      });
    } catch (error) {
      ctx.trace.add({
        stepId: step.id,
        operation: "REPAIR",
        resolvedArgs: {
          targetVar: step.targetVar,
          reportVar: step.reportVar,
          max: step.maxIterations
        },
        outputSummary: "Repair loop failed",
        status: "fail",
        durationMs: Date.now() - started,
        errors: [error instanceof Error ? error.message : String(error)]
      });
      throw error;
    }
  }

  private resolveArgs(args: Record<string, GlyphIRValue>, ctx: RuntimeContext, stepId: string): Record<string, unknown> {
    return Object.fromEntries(Object.entries(args).map(([key, value]) => [key, this.resolveValue(value, ctx, stepId)]));
  }

  private resolveValue(value: GlyphIRValue, ctx: RuntimeContext, stepId: string): unknown {
    if (Array.isArray(value)) {
      return value.map((item) => this.resolveValue(item, ctx, stepId));
    }

    if (value !== null && typeof value === "object") {
      const keys = Object.keys(value);

      if (keys.length === 1 && "var" in value && typeof value.var === "string") {
        return this.requireVariable(value.var, ctx, stepId);
      }

      if (keys.length === 1 && "ctx" in value && typeof value.ctx === "string") {
        return this.resolveContext(value.ctx, ctx, stepId);
      }

      return Object.fromEntries(Object.entries(value).map(([key, nested]) => [key, this.resolveValue(nested, ctx, stepId)]));
    }

    return value;
  }

  private requireVariable(name: string, ctx: RuntimeContext, stepId: string): unknown {
    if (!ctx.variables.has(name)) {
      throw new GlyphRuntimeError(`Unknown variable "${name}"`, stepId);
    }

    return ctx.variables.get(name);
  }

  private resolveContext(path: string, ctx: RuntimeContext, stepId: string): unknown {
    const parts = path.split(".");
    let current: unknown = ctx.context;

    for (const part of parts) {
      if (current === null || typeof current !== "object" || !(part in current)) {
        throw new GlyphRuntimeError(`Unknown ctx reference "ctx.${path}"`, stepId);
      }

      current = (current as Record<string, unknown>)[part];
    }

    return current;
  }

  private reportStatus(value: unknown): ToolStatus {
    if (value !== null && typeof value === "object" && "status" in value) {
      const status = (value as { status?: unknown }).status;
      if (status === "pass" || status === "warning" || status === "fail") {
        return status;
      }
    }

    return "warning";
  }
}

File: src/runtime/glyph_vm.rs
use std::collections::BTreeMap;
use std::time::Instant;

use serde::Serialize;
use serde_json::{Map, Value};

use crate::harness::tool_registry::ToolRegistry;
use crate::harness::types::ToolStatus;
use crate::ir::glyph_ir::{
    GlyphIr, GlyphIrStep, GlyphRepairStep, GlyphToolStep, parse_glyph_to_ir,
};
use crate::ir::validate_ir::{GlyphIrValidationError, validate_ir};

use super::context::RuntimeContext;
use super::errors::GlyphRuntimeError;
use super::trace::TraceEvent;

#[derive(Debug, Clone, Serialize)]
pub struct GlyphVmRunResult {
    pub ir: GlyphIr,
    pub trace: Vec<TraceEvent>,
    pub outputs: Vec<Value>,
    pub variables: BTreeMap<String, Value>,
    #[serde(rename = "mockFS")]
    pub mock_fs: BTreeMap<String, Value>,
}

#[derive(Debug, Clone, Default)]
pub struct GlyphVmOptions {
    pub mock_fs: BTreeMap<String, Value>,
}

#[derive(Debug, thiserror::Error)]
pub enum GlyphVmError {
    #[error(transparent)]
    Syntax(#[from] crate::language::errors::GlyphSyntaxError),
    #[error(transparent)]
    Validation(#[from] GlyphIrValidationError),
    #[error(transparent)]
    Runtime(#[from] GlyphRuntimeError),
}

pub struct GlyphVm {
    registry: ToolRegistry,
}

impl GlyphVm {
    pub fn new(registry: ToolRegistry) -> Self {
        Self { registry }
    }

    pub fn run_source(&self, source: &str) -> Result<GlyphVmRunResult, GlyphVmError> {
        let ir = validate_ir(parse_glyph_to_ir(source)?)?;
        self.execute(ir, GlyphVmOptions::default())
    }

    pub fn execute(
        &self,
        ir: GlyphIr,
        options: GlyphVmOptions,
    ) -> Result<GlyphVmRunResult, GlyphVmError> {
        let ir = validate_ir(ir)?;
        let mut ctx = RuntimeContext::new(ir.context.clone(), options.mock_fs);

        for flow in &ir.flows {
            for step in &flow.steps {
                self.execute_step(step, &mut ctx, None)?;
            }
        }

        Ok(GlyphVmRunResult {
            ir,
            trace: ctx.trace.all(),
            outputs: ctx.outputs.clone(),
            variables: ctx.variables.snapshot(),
            mock_fs: ctx.snapshot_fs(),
        })
    }

    fn execute_step(
        &self,
        step: &GlyphIrStep,
        ctx: &mut RuntimeContext,
        iteration: Option<usize>,
    ) -> Result<(), GlyphRuntimeError> {
        match step {
            GlyphIrStep::Tool(tool) => self.execute_tool_step(tool, ctx, iteration),
            GlyphIrStep::Repair(repair) => self.execute_repair_step(repair, ctx),
        }
    }

    fn execute_tool_step(
        &self,
        step: &GlyphToolStep,
        ctx: &mut RuntimeContext,
        iteration: Option<usize>,
    ) -> Result<(), GlyphRuntimeError> {
        let started = Instant::now();
        let mut resolved_args = Map::new();

        let result: Result<(), GlyphRuntimeError> = (|| {
            resolved_args = self.resolve_args(&step.args, ctx, &step.id)?;
            let tool = self.registry.get(&step.op).ok_or_else(|| {
                GlyphRuntimeError::new(
                    format!("Unknown tool \"{}\"", step.op),
                    Some(step.id.clone()),
                )
            })?;
            let result = tool(resolved_args.clone(), ctx)?;

            if let Some(assign_to) = &step.assign_to {
                ctx.variables.set(assign_to, result.value.clone());
            }

            if step.op == "EXPORT" {
                ctx.outputs.push(result.value.clone());
            }

            ctx.trace.add(TraceEvent {
                step_id: step.id.clone(),
                operation: step.op.clone(),
                resolved_args: resolved_args.clone(),
                output_summary: result.summary.clone(),
                status: result.status,
                duration_ms: started.elapsed().as_millis(),
                errors: if result.status == ToolStatus::Fail {
                    result.warnings.clone()
                } else {
                    vec![]
                },
                iteration,
            });

            Ok(())
        })();

        if let Err(error) = result {
            ctx.trace.add(TraceEvent {
                step_id: step.id.clone(),
                operation: step.op.clone(),
                resolved_args,
                output_summary: "Step failed".to_string(),
                status: ToolStatus::Fail,
                duration_ms: started.elapsed().as_millis(),
                errors: vec![error.to_string()],
                iteration,
            });
            return Err(error);
        }

        Ok(())
    }

    fn execute_repair_step(
        &self,
        step: &GlyphRepairStep,
        ctx: &mut RuntimeContext,
    ) -> Result<(), GlyphRuntimeError> {
        let started = Instant::now();

        let result: Result<(), GlyphRuntimeError> = (|| {
            self.require_variable(&step.target_var, ctx, &step.id)?;
            self.require_variable(&step.report_var, ctx, &step.id)?;

            let mut iterations = 0usize;
            for index in 0..step.max_iterations {
                if self.report_status(ctx.variables.get(&step.report_var)) == ToolStatus::Pass {
                    break;
                }

                iterations = index + 1;
                for inner in &step.steps {
                    self.execute_step(inner, ctx, Some(iterations))?;
                }

                if self.report_status(ctx.variables.get(&step.report_var)) == ToolStatus::Pass {
                    break;
                }
            }

            let final_status = self.report_status(ctx.variables.get(&step.report_var));
            ctx.trace.add(TraceEvent {
                step_id: step.id.clone(),
                operation: "REPAIR".to_string(),
                resolved_args: Map::from_iter([
                    (
                        "target".to_string(),
                        ctx.variables
                            .get(&step.target_var)
                            .cloned()
                            .unwrap_or(Value::Null),
                    ),
                    (
                        "report".to_string(),
                        ctx.variables
                            .get(&step.report_var)
                            .cloned()
                            .unwrap_or(Value::Null),
                    ),
                    ("max".to_string(), Value::from(step.max_iterations)),
                ]),
                output_summary: format!(
                    "Repair loop completed after {iterations} iteration{}",
                    if iterations == 1 { "" } else { "s" }
                ),
                status: final_status,
                duration_ms: started.elapsed().as_millis(),
                errors: vec![],
                iteration: None,
            });

            Ok(())
        })();

        if let Err(error) = result {
            ctx.trace.add(TraceEvent {
                step_id: step.id.clone(),
                operation: "REPAIR".to_string(),
                resolved_args: Map::from_iter([
                    (
                        "targetVar".to_string(),
                        Value::String(step.target_var.clone()),
                    ),
                    (
                        "reportVar".to_string(),
                        Value::String(step.report_var.clone()),
                    ),
                    ("max".to_string(), Value::from(step.max_iterations)),
                ]),
                output_summary: "Repair loop failed".to_string(),
                status: ToolStatus::Fail,
                duration_ms: started.elapsed().as_millis(),
                errors: vec![error.to_string()],
                iteration: None,
            });
            return Err(error);
        }

        Ok(())
    }

    fn resolve_args(
        &self,
        args: &Map<String, Value>,
        ctx: &RuntimeContext,
        step_id: &str,
    ) -> Result<Map<String, Value>, GlyphRuntimeError> {
        args.iter()
            .map(|(key, value)| Ok((key.clone(), self.resolve_value(value, ctx, step_id)?)))
            .collect()
    }

    fn resolve_value(
        &self,
        value: &Value,
        ctx: &RuntimeContext,
        step_id: &str,
    ) -> Result<Value, GlyphRuntimeError> {
        match value {
            Value::Array(items) => items
                .iter()
                .map(|item| self.resolve_value(item, ctx, step_id))
                .collect::<Result<Vec<_>, _>>()
                .map(Value::Array),
            Value::Object(object) => {
                if object.len() == 1
                    && let Some(Value::String(name)) = object.get("var")
                {
                    return self.require_variable(name, ctx, step_id).cloned();
                }

                if object.len() == 1
                    && let Some(Value::String(path)) = object.get("ctx")
                {
                    return self.resolve_context(path, ctx, step_id);
                }

                object
                    .iter()
                    .map(|(key, nested)| {
                        Ok((key.clone(), self.resolve_value(nested, ctx, step_id)?))
                    })
                    .collect::<Result<Map<_, _>, _>>()
                    .map(Value::Object)
            }
            _ => Ok(value.clone()),
        }
    }

    fn require_variable<'a>(
        &self,
        name: &str,
        ctx: &'a RuntimeContext,
        step_id: &str,
    ) -> Result<&'a Value, GlyphRuntimeError> {
        ctx.variables.get(name).ok_or_else(|| {
            GlyphRuntimeError::new(
                format!("Unknown variable \"{name}\""),
                Some(step_id.to_string()),
            )
        })
    }

    fn resolve_context(
        &self,
        path: &str,
        ctx: &RuntimeContext,
        step_id: &str,
    ) -> Result<Value, GlyphRuntimeError> {
        let mut current = Value::Object(ctx.context.clone());

        for part in path.split('.') {
            match current {
                Value::Object(object) => {
                    current = object.get(part).cloned().ok_or_else(|| {
                        GlyphRuntimeError::new(
                            format!("Unknown ctx reference \"ctx.{path}\""),
                            Some(step_id.to_string()),
                        )
                    })?;
                }
                _ => {
                    return Err(GlyphRuntimeError::new(
                        format!("Unknown ctx reference \"ctx.{path}\""),
                        Some(step_id.to_string()),
                    ));
                }
            }
        }

        Ok(current)
    }

    fn report_status(&self, value: Option<&Value>) -> ToolStatus {
        value
            .and_then(|value| value.get("status"))
            .and_then(Value::as_str)
            .map(|status| match status {
                "pass" => ToolStatus::Pass,
                "warning" => ToolStatus::Warning,
                "fail" => ToolStatus::Fail,
                _ => ToolStatus::Warning,
            })
            .unwrap_or(ToolStatus::Warning)
    }
}
